In [1]:
import os
import sys
import glob
import time
import pandas as pd
import json
import datetime

# Data Paths

In [2]:
data_path = "/DATA/7A/xblock/"
df_path = "/DATA/7A/assets/dataframes/"

#
erc20_transactions_path = "ERC20Transactions/uncompressed"
erc721_transactions_path = "ERC721Transactions/uncompressed"
internal_transactions_path = "InternalTransactions/uncompressed"
normal_transactions_path = "NormalTransactions/uncompressed"

#token information
token_info_path = "TokenInfo/uncompressed/"
erc20_filename = "ERC20TokenInfo.csv"
erc721_filename = "ERC721TokenInfo.csv"

#dataframes
internalTransactions_filename = "internalTransactions.df"
erc20Transactions_filename = "erc20Transactions.df"
erc721Transactions_filename = "erc721Transactions.df"


# Get Token Information and Transactions

##  get erc721 token information

In [3]:
#get erc721 token information
erc721InfoDF = pd.read_csv(data_path + token_info_path + erc721_filename, error_bad_lines=False)
erc721InfoDF = erc721InfoDF[erc721InfoDF['decimal'] != "None"].astype({"decimal": float})
print(erc721InfoDF.shape[0])
erc721InfoDF.head()

7691


address              name  symbol  \
0  0x27b4bc90fbe56f02ef50f2e2f79d7813aa8941a7         gego.dego    GEGO   
1  0x3f9d7b15763f57d7f5015901eaccca27ce996bb2        Polysphere  PLYSPR   
2  0x06012c8cf97bead5deae237070f9587f8e7a266d     CryptoKitties      CK   
3  0xbbe7aeb729e65e3854a30d583cee6a28434eb841        Polysphere    POLY   
4  0x2fb5d7dda4f1f20f974a0fdd547c38674e8d940c  KnightStory Item    MyKI   

     decimal  
0    54084.0  
1        5.0  
2  1996597.0  
3        0.0  
4    13250.0

##  get erc20 token information

In [4]:
#get erc20 token information
#line 154118 in this file is problematic for some reason
erc20InfoDF = pd.read_csv(data_path + token_info_path + erc20_filename, error_bad_lines=False)
print(erc20InfoDF.shape[0])
erc20InfoDF.head()

226250


b'Skipping line 154118: expected 5 fields, saw 7\n'


address                  name symbol  \
0  0x6b0359f95796327475ad4f12ae4e1047c3a67fa3         FinFine Token    FFT   
1  0x12d79c345cac7b050a5ff0797b5a607e254c73f5                Buytex    BUX   
2  0x6bc1f3a1ae56231dbb64d3e82e070857eae86045                xensor    XSR   
3  0x10086399dd8c1e3de736724af52587a2044c9fa2  The Midas Touch Gold   TMTG   
4  0xdac17f958d2ee523a2206206994597c13d831ec7            Tether USD   USDT   

                    totalSupply decimal  
0                  214285500000       3  
1                 9999999999999       4  
2  4800000000000000000000000000      18  
3  9161139666073037500000000000      18  
4             20817977739617189       6

## read all internal transactions

In [4]:
#all_files = glob.glob(data_path + internal_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end='', flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#internalTransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
internalTransactionsDF = pd.read_feather(df_path + internalTransactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(internalTransactionsDF.shape[0])
internalTransactionsDF.head(10)

NameError: name 'df_path' is not defined

## read all erc20 transactions

In [5]:
#all_files = glob.glob(data_path + erc20_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end='', flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#erc20TransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
erc20TransactionsDF = pd.read_feather(df_path + erc20Transactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(erc20TransactionsDF.shape[0])
erc20TransactionsDF.head()

Took 660.3593909740448  seconds.
601955094


blockNumber   timestamp                                    transactionHash  \
0       462680  1446203330  0x806172da2073b5c3d961a8c613b7e469766ad19c6220...   
1       462683  1446203369  0xd5c17b09dcf6e5daa0566d7d2a8245f10a755ecfbf34...   
2       462696  1446203551  0x530e047cccef3ea1f78bcdb629661fabe193ba87cd82...   
3       477958  1446458222  0x0d58cd643c38a8951ee26a49399d4d5682fab74e96a9...   
4       483333  1446552343  0xd628780ba231cefe6a4f6c3da3b683b16f6151dc9753...   

                                 tokenAddress  \
0  0xe6ee69495b571e1042f760d7f34009164aff87a2   
1  0xe6ee69495b571e1042f760d7f34009164aff87a2   
2  0xe6ee69495b571e1042f760d7f34009164aff87a2   
3  0xe6ee69495b571e1042f760d7f34009164aff87a2   
4  0xf4eced2f682ce333f96f2d8966c613ded8fc95dd   

                                         from  \
0  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
1  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
2  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
3  0xb1a2b43a7433dd150bb82227ed519cd6b142d382   
4  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f   

                                           to  fromIsContract  toIsContract  
0  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f               0             0  
1  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f               0             0  
2  0x9b22a80d5c7b3374a05b446081f97d0a34079e7f               0             0  
3  0xcf8470d3388a198418e02aaf8e37392cfbc18867               0             1  
4  0xd1220a0cf47c7b9be7a2e6ba89f429762e7b9adb               0             0

## read all erc721 transactions

In [7]:
#all_files = glob.glob(data_path + erc721_transactions_path + "/*.csv")
#
#li = []
#
#for filename in all_files:
#    start = time.time()
#    print("Reading file", filename, "...", end="", flush=True)
#    df = pd.read_csv(filename, index_col=None, header=0)
#    li.append(df)
#    end = time.time()
#    print(" took", end - start, "seconds.")
#    #break #read just the first file
#erc721TransactionsDF = pd.concat(li, axis=0, ignore_index=True)

s = time.time()
erc721TransactionsDF = pd.read_feather(df_path + erc721Transactions_filename)
e = time.time()
print("Took", e - s, " seconds.")
print(erc721TransactionsDF.shape[0])
erc721TransactionsDF.head()

Took 12.7132728099823  seconds.
29634892


blockNumber   timestamp                                    transactionHash  \
0      3006260  1484564580  0x9d5fe398fe902975f8508e197ec6c0d0082f8a142a7f...   
1      3053922  1485244952  0xd97c4ddc0f5c0af8c18a093b184de0a8a14e35c87c3f...   
2      3053922  1485244952  0xd97c4ddc0f5c0af8c18a093b184de0a8a14e35c87c3f...   
3      3053922  1485244952  0xd97c4ddc0f5c0af8c18a093b184de0a8a14e35c87c3f...   
4      3132454  1486369652  0x2a7ed459141cc841ab5b376ea64dbc7041b333384d7a...   

                                 tokenAddress  \
0  0xe9436e4cb00f38215a941c7141ee98348fdcce6d   
1  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   
2  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   
3  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   
4  0x55b9a11c2e8351b4ffc7b11561148bfac9977855   

                                         from  \
0  0x002add5778a3d327278a357a4983a57c1894af12   
1  0x121b1badeda482fa25f01c63ad17bbc2701f9165   
2  0x121b1badeda482fa25f01c63ad17bbc2701f9165   
3  0x121b1badeda482fa25f01c63ad17bbc2701f9165   
4  0x01d80e458ae6cd5d2bce90b06ff2d3fa993b65f5   

                                           to  fromIsContract  toIsContract  
0  0xa84ab927943d87e2f14afac62af7283e26cf010e               0             0  
1  0x5d2c24efac49825f682b5d334428f1a5c96acc09               0             1  
2  0x3effc3bb636a66d39ae016399d787d99e1c286a4               0             0  
3  0x5d2c24efac49825f682b5d334428f1a5c96acc09               0             1  
4  0x5d2c24efac49825f682b5d334428f1a5c96acc09               0             1

***

***

***

# Analysis

In [8]:
import numpy as np

## Token Statistics

In [9]:
targetInfoDF = erc20InfoDF
targetDF = erc20TransactionsDF


s = time.time()
#get contract addresses
contractAdresses = targetDF[targetDF["tokenAddress"].isin(targetInfoDF["address"].values)]

#count transactions by tokenAddress 
contractAdresses = contractAdresses.tokenAddress.value_counts()
targetInfoDF['number_of transactions'] = targetInfoDF['address'].map(contractAdresses)
targetInfoDF = targetInfoDF.sort_values(by=['number_of transactions'], ascending=False, na_position='last')
e = time.time()
print("Took", e - s, " seconds.")

#token_address = "0xD1CEeeeee83F8bCF3BEDad437202b6154E9F5405"
#token_symbol = "CHZ"
#print(targetInfoDf[targetInfoDf["symbol"] == token_symbol])
#print(targetInfoDf[targetInfoDf["address"] == token_address])
targetInfoDF.head(30)

Took 166.4360637664795  seconds.


address                   name  \
4      0xdac17f958d2ee523a2206206994597c13d831ec7             Tether USD   
5      0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2          Wrapped Ether   
22     0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48               USD Coin   
4620   0x174bfa6600bf90c885c7c01c7031389ed1461ab9         More Gold Coin   
50     0x6b175474e89094c44da98b954eedeac495271d0f         Dai Stablecoin   
37     0x514910771af9ca656af840dff83e8264ecf986ca        ChainLink Token   
1005   0x58b6a8a3302369daec383334672404ee733ab239         Livepeer Token   
525    0x03cb0021808442ad5efb61197966aef72a1def96                coToken   
79946  0x86fa049857e0209aa7d9e616f7eb3b3b78ecfdb0                    NaN   
561    0x1f573d6fb3f13d689ff844b4ce37794d79a7ff1c   Bancor Network Token   
915    0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359                   None   
576    0x8e870d67f660d95d5be530380d0ec0bd388289e1         Paxos Standard   
444    0xd26114cd6ee289accf82350c8d8487fedb8a0c07               OMGToken   
394    0x0d8775f648430679a709e98d2b0cb6250d2887ef  Basic Attention Token   
3602   0xf230b790e05390fc8295f4d3f60332c93bed42e2                 Tronix   
7665   0xe62e6e6c3b808faad3a54b226379466544d76ea4           Liberty Cash   
3287   0x8fdcc30eda7e94f1c12ce0280df6cd531e8365c5               CpcToken   
9418   0xc0829421c1d260bd3cb3e0f06cfe2d52db2ce315            Ether Token   
2169   0xac08809df1048b82959d6251fbc9538920bed1fa                    MSD   
94     0x1f9840a85d5af5bf1d1762f925bdaddc4201f984                Uniswap   
509    0xe41d2489571d322189246dafa5ebde1f4699f498      0x Protocol Token   
399    0xb64ef51c888972c908cfacf59b47c1afbc0ab8ac             StorjToken   
128    0x0000000000004946c0e9f43f4dee607b0ef1fa1c  Chi Gastoken by 1inch   
30683  0xdc17e8a84fee8b52e4de7a85160f8cdbb3bb2494           BitAnex Coin   
64     0x6b3595068778dd592e39a122f4f5a5cf09c90fe2             SushiToken   
6883   0x8e766f57f7d16ca50b4a0b90b88f6468a09b0439          Maximine Coin   
2840   0xa821f14fb6394e82839f5161f214cacc90372453             Esch$Token   
7036   0x74fd51a98a4a1ecbef8cc43be801cce630e260bd            SiaCashCoin   
71307  0x9e15f8ad98e95033c1d4798458cec34a4b5972b0           Alchemy Coin   
171    0x2b591e99afe9f32eaa6214f7b7629768c40eeb39                    HEX   

        symbol                    totalSupply decimal  number_of transactions  
4         USDT              20817977739617189       6              93669514.0  
5         WETH      7370101322099794115939926      18              36038706.0  
22        USDC               9215344513959550       6              13084473.0  
4620       MGC    600000000000000000000000000      18               8947534.0  
50         DAI   2562847855060761174347642812      18               8079570.0  
37        LINK   1000000000000000000000000000      18               7651947.0  
1005       LPT     23306762282308177576176346      18               5638056.0  
525    coToken         1114753470684460878279      18               5367786.0  
79946     None   1000000000000000000000000000      18               3570224.0  
561        BNT    158117591896340222167262079      18               3321274.0  
915       None      3729494392346553152988922      18               3151049.0  
576        PAX    760384582690000000000000000      18               3103029.0  
444        OMG    140245398245132780789239631      18               3093435.0  
394        BAT   1500000000000000000000000000      18               2976051.0  
3602       TRX             100000000000000000       6               2935695.0  
7665       LCS                567108520655905       6               2790275.0  
3287      CPCT   2000000000000000000000000000      18               2220879.0  
9418       ETH           32039234672376792014      18               2208994.0  
2169       MSD           10000000000000000008       8               2113115.0  
94         UNI   1000000000000000000000000000      18               1980131.0  
509  

Some erc721 tokens have a different decimal and number of transactions. Why is that?

**Different**

CryptoKitties => decimal 1996597, number_of transactions 5434740

MARBLE-NFT => decimal 60452, number of transactions 187361

**Same**

Gods Unchained Cards => decimal 6680552, number_of transactions 6680552

Gods Unchained => decimal 603050, res 603050

Gods Unchained Cards => decimal 580736, res 580736



**Dapps by Type**

Games:
https://sorare.com/
https://cryptodozer.io/
https://godsunchained.com/

Marketplace:
https://rarible.com/

### I believe we need relatively recent data ! (Not sure tho)



## Significance of tokens

In [16]:
erc721InfoDF.head(10)
erc20InfoDF.head(10)

erc721Addresses = erc721InfoDF["address"].values
erc20Addresses = erc20InfoDF["address"].values
tokenAddresses = np.concatenate([erc20Addresses, erc721Addresses])
print(len(tokenAddresses))
tokenAddresses

# token transcation ratio
tokenTransactionRatio = 

233941


array(['0x6b0359f95796327475ad4f12ae4e1047c3a67fa3',
       '0x12d79c345cac7b050a5ff0797b5a607e254c73f5',
       '0x6bc1f3a1ae56231dbb64d3e82e070857eae86045', ...,
       '0xf5c8380f5656f7d82b36dfe5fa2f484839e2fc39',
       '0x6721f071cf1a7d10f5309a896ffa1a724a9a705c',
       '0xf74db21536d2ee9b8428bfee0394586771e13bf0'], dtype=object)